extension-iap/src/java/com/defold/iap/IapGooglePlay.java

```diff
@@ -5,507 +5,382 @@
-        this.handler = new Handler(this);
-        this.messenger = new Messenger(this.handler);
-
-        serviceConn = new ServiceConnection() {
 
+        billingClient = BillingClient.newBuilder(activity).setListener(this).enablePendingPurchases().build();
+        billingClient.startConnection(new BillingClientStateListener() {
             @Override
-            public void onServiceDisconnected(ComponentName name) {
-                Log.v(TAG, "IAP disconnected");
-                service = null;
+            public void onBillingSetupFinished(BillingResult billingResult) {
+                if (billingResult.getResponseCode() == BillingResponseCode.OK) {
+                    Log.v(TAG, "Setup finished");
+                    // NOTE: we will not query purchases here. This is done
+                    // when the extension listener is set
+                }
+                else {
+                    Log.wtf(TAG, "Setup error: " + billingResult.getDebugMessage());
+                }
             }
 
             @Override
-            public void onServiceConnected(ComponentName name, IBinder binderService) {
-                Log.v(TAG, "IAP connected");
-                service = IInAppBillingService.Stub.asInterface(binderService);
-                skuDetailsThread = new SkuDetailsThread();
-                skuDetailsThread.start();
+            public void onBillingServiceDisconnected() {
+                Log.v(TAG, "Service disconnected");
             }
-        };
-
-        Intent serviceIntent = new Intent("com.android.vending.billing.InAppBillingService.BIND");
-        // Limit intent to vending package
-        serviceIntent.setPackage("com.android.vending");
-        List<ResolveInfo> intentServices = activity.getPackageManager().queryIntentServices(serviceIntent, 0);
-        if (intentServices != null && !intentServices.isEmpty()) {
-            // service available to handle that Intent
-            activity.bindService(serviceIntent, serviceConn, Context.BIND_AUTO_CREATE);
-        } else {
-            serviceConn = null;
-            Log.e(TAG, "Billing service unavailable on device.");
-        }
+        });
     }
 
     public void stop() {
-        this.activity.runOnUiThread(new Runnable() {
-            @Override
-            public void run() {
-                if (serviceConn != null) {
-                    activity.unbindService(serviceConn);
-                    serviceConn = null;
-                }
-                if (skuDetailsThread != null) {
-                    skuDetailsThread.stop = true;
-                    skuDetailsThread.interrupt();
-                    try {
-                        skuDetailsThread.join();
-                    } catch (InterruptedException e) {
-                        Log.wtf(TAG, "Failed to join thread", e);
-                    }
-                }
-            }
-        });
+        Log.d(TAG, "stop()");
+        billingClient.endConnection();
```

Comment:
Comparing with the old code, it was "secured" from being called from within a callback, which might come from another thread.

Is this no longer the case?

Reply:
I thought that wouldn't be a problem anymore, but the documentation isn't clear and their example is a bit bloated an uses a lifecycle manager instance: https://github.com/android/play-billing-samples/blob/master/ClassyTaxiJava/app/src/main/java/com/example/android/classytaxijava/billing/BillingClientLifecycle.java#L113